File: migrations/2026-01-28/lib/schemas/v1/customer.rb
# migrations/2026-01-28/lib/schemas/v1/customer.rb
#
# frozen_string_literal: true

require_relative '../base'

module Migration
  module Schemas
    module V1
      # JSON Schema for V1 customer data (input from Redis HGETALL).
      #
      # V1 customers use email as custid and store fields as Redis hash.
      # All values are strings since Redis hashes only store strings.
      #
      CUSTOMER = {
        '$schema' => 'http://json-schema.org/draft-07/schema#',
        'title' => 'Customer V1',
        'description' => 'V1 customer record from Redis hash (pre-migration)',
        'type' => 'object',
        'required' => %w[custid],
        'properties' => {
          # Primary identifier (email in V1)
          'custid' => {
            'type' => 'string',
            'minLength' => 1,
            'description' => 'Customer identifier (email address in V1)',
          },

          # Optional email (may be absent if custid is email)
          'email' => {
            'type' => 'string',
            'description' => 'Email address',
          },

          # Timestamps (stored as epoch floats as strings)
          'created' => {
            'type' => 'string',
            'pattern' => '^\\d+(\\.\\d+)?$',
            'description' => 'Creation timestamp (epoch float as string)',
          },
          'updated' => {
            'type' => 'string',
            'pattern' => '^\\d+(\\.\\d+)?$',
            'description' => 'Last update timestamp (epoch float as string)',
          },

          # Role (customer, colonel, recipient, anonymous)
          'role' => {
            'type' => 'string',
            'enum' => %w[customer colonel recipient anonymous],
            'description' => 'User role',
          },

          # Verification status (stored as string boolean or 0/1)
          'verified' => {
            'type' => 'string',
            'enum' => ['true', 'false', '0', '1', ''],
            'description' => 'Email verification status',
          },

          # Plan identifier
          'planid' => {
            'type' => 'string',
            'description' => 'Subscription plan identifier',
          },

          # Stripe customer ID (empty string or cus_*)
          'stripe_customer_id' => {
            'type' => 'string',
            'pattern' => '^(cus_.*)?$',
            'description' => 'Stripe customer identifier',
          },

          # Stripe subscription ID (empty string or sub_*)
          'stripe_subscription_id' => {
            'type' => 'string',
            'pattern' => '^(sub_.*)?$',
            'description' => 'Stripe subscription identifier',
          },

          # Locale
          'locale' => {
            'type' => 'string',
            'description' => 'User locale preference',
          },

          # API token
          'apitoken' => {
            'type' => 'string',
            'description' => 'API authentication token',
          },

          # Passphrase (hashed)
          'passphrase' => {
            'type' => 'string',
            'description' => 'Hashed authentication passphrase',
          },

          # Last login timestamp (empty string or epoch float)
          'last_login' => {
            'type' => 'string',
            'pattern' => '^(\\d+(\\.\\d+)?)?$',
            'description' => 'Last login timestamp (epoch float as string)',
          },

          # Secret count
          'secrets_created' => {
            'type' => 'string',
            'pattern' => '^\\d+$',
            'description' => 'Total secrets created (integer as string)',
          },

          # Active status
          'active' => {
            'type' => 'string',
            'enum' => ['true', 'false', '0', '1', ''],
            'description' => 'Account active status',
          },

          # Contributor flag
          'contributor' => {
            'type' => 'string',
            'enum' => ['true', 'false', '0', '1', ''],
            'description' => 'Contributor flag',
          },

          # Counter: emails sent
          'emails_sent' => {
            'type' => 'string',
            'pattern' => '^\\d+$',
            'description' => 'Number of emails sent (integer as string)',
          },

          # Duplicate of custid (deprecated)
          'key' => {
            'type' => 'string',
            'description' => 'Duplicate of custid (deprecated)',
          },

          # Hash algorithm indicator (1=bcrypt, 2=argon2)
          'passphrase_encryption' => {
            'type' => 'string',
            'enum' => %w[1 2],
            'description' => 'Passphrase hash algorithm (1=bcrypt, 2=argon2)',
          },

          # Counter: secrets burned
          'secrets_burned' => {
            'type' => 'string',
            'pattern' => '^\\d+$',
            'description' => 'Number of secrets burned (integer as string)',
          },

          # Counter: secrets shared
          'secrets_shared' => {
            'type' => 'string',
            'pattern' => '^\\d+$',
            'description' => 'Number of secrets shared (integer as string)',
          },

          # Session ID (deprecated)
          'sessid' => {
            'type' => 'string',
            'description' => 'Session ID (deprecated)',
          },

          # Stripe checkout email
          'stripe_checkout_email' => {
            'type' => 'string',
            'description' => 'Email used in Stripe checkout',
          },
        },
        'additionalProperties' => true,
      }.freeze

      # Register the schema
      Schemas.register(:customer_v1, CUSTOMER)
    end
  end
end

__END__

## V1 Dump Field Analysis (2026-01-29)

**Source:** `exports/customer/customer_dump.jsonl`
**Total Records:** 400 customer objects

### V1 Customer Fields (22 fields)

| Field | Description | Coverage |
|-------|-------------|----------|
| `apitoken` | API authentication token | 41 with values |
| `contributor` | Contributor flag | 0 with values |
| `created` | Account creation timestamp (epoch) | 400 |
| `custid` | Customer ID (email address in v1) | 400 |
| `email` | Email address | 400 |
| `emails_sent` | Counter: emails sent | 400 |
| `key` | Duplicate of custid | 400 |
| `last_login` | Last login timestamp | 400 |
| `locale` | User locale preference | 400 |
| `passphrase` | Hashed password (bcrypt) | 397 |
| `passphrase_encryption` | Hash algorithm (1=bcrypt, 2=argon2) | 398 (all bcrypt) |
| `planid` | Subscription plan ID | 400 |
| `role` | User role (customer, etc.) | 400 |
| `secrets_burned` | Counter: secrets burned | 400 |
| `secrets_created` | Counter: secrets created | 400 |
| `secrets_shared` | Counter: secrets shared | 400 |
| `sessid` | Session ID (deprecated) | 400 |
| `stripe_checkout_email` | Stripe checkout email | 0 with values |
| `stripe_customer_id` | Stripe customer ID | 12 with values |
| `stripe_subscription_id` | Stripe subscription ID | 12 with values |
| `updated` | Last update timestamp | 400 |
| `verified` | Account verification status | 400 |

### V1 Schema Coverage

**Schema has 15 fields. Dump has 22 fields.**

**In schema (15):** `custid`, `email`, `created`, `updated`, `role`, `verified`, `planid`, `stripe_customer_id`, `stripe_subscription_id`, `locale`, `apitoken`, `passphrase`, `last_login`, `secrets_created`, `active`

**Missing from schema (7):**
- `contributor` - Contributor flag
- `emails_sent` - Counter: emails sent
- `key` - Duplicate of custid (deprecated)
- `passphrase_encryption` - Hash algorithm indicator
- `secrets_burned` - Counter: secrets burned
- `secrets_shared` - Counter: secrets shared
- `sessid` - Session ID (deprecated)
- `stripe_checkout_email` - Stripe checkout email

File: src/schemas/contracts/customer.ts
// src/schemas/contracts/customer.ts
// @see src/tests/contracts/customer-schema-contract.spec.ts - Contract tests
//
// Canonical customer record schema - field names and output types only.
// Version-specific schemas (V2, V3) extend this with wire-format transforms.
//
// This schema owns the field contract. V2/V3 own the encoding.

/**
 * Customer record contracts defining field names and output types.
 *
 * Customers represent authenticated users in the system. These canonical
 * schemas define the "what" (field names and final types) without the "how"
 * (wire-format transforms).
 *
 * Version-specific shapes in `shapes/v2/customer.ts` and `shapes/v3/customer.ts`
 * extend these with appropriate transforms for each API version.
 *
 * @module contracts/customer
 * @category Contracts
 * @see {@link "shapes/v2/customer"} - V2 wire format with string transforms
 * @see {@link "shapes/v3/customer"} - V3 wire format with native types
 */

import { z } from 'zod';

// ─────────────────────────────────────────────────────────────────────────────
// Role enum and values
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Customer role values as a const tuple.
 *
 * Roles determine authorization level and user status:
 * - `customer`: Standard authenticated user
 * - `colonel`: Administrative/privileged user
 * - `admin`: Administrative user (assignable via `bin/ots customers role`)
 * - `staff`: Staff user (assignable via `bin/ots customers role`)
 * - `recipient`: Read-only secret recipient
 * - `user_deleted_self`: Self-deleted account (soft delete)
 * - `anonymous`: Unauthenticated sentinel (`Customer#anonymous?` backend check)
 *
 * MUST be a superset of the backend's assignable roles —
 * `Auth::Operations::Customers::SetRole::VALID_ROLES` in
 * apps/web/auth/operations/customers/set_role.rb. A backend-writable role
 * missing here broke /account/settings/api for promoted accounts (#4298,
 * FRONTEND-19V): the strict enum failed the whole AccountResponse parse.
 * The sync is enforced by src/tests/contracts/customer-role-contract.spec.ts,
 * which reads VALID_ROLES from the Ruby source.
 *
 * @category Contracts
 * @example
 * ```typescript
 * // Use with Zod enum
 * const roleSchema = z.enum(customerRoleValues);
 *
 * // Type narrowing
 * if (customerRoleValues.includes(value as CustomerRole)) {
 *   // value is CustomerRole
 * }
 * ```
 */
export const customerRoleValues = [
  'customer',
  'colonel',
  'admin',
  'staff',
  'recipient',
  'user_deleted_self',
  'anonymous',
] as const;

export type CustomerRole = (typeof customerRoleValues)[number];

/**
 * Customer role enum object for runtime role checks.
 *
 * Using const object pattern over enum because:
 * 1. Produces simpler runtime code (just a plain object vs IIFE)
 * 2. Better tree-shaking since values can be inlined
 * 3. Works naturally with Zod's z.enum()
 *
 * @category Contracts
 * @example
 * ```typescript
 * if (customer.role === CustomerRole.COLONEL) {
 *   // User has admin privileges
 * }
 *
 * // Use in switch statements
 * switch (customer.role) {
 *   case CustomerRole.CUSTOMER:
 *     return 'Standard User';
 *   case CustomerRole.COLONEL:
 *     return 'Administrator';
 *   case CustomerRole.RECIPIENT:
 *     return 'Recipient';
 *   case CustomerRole.USER_DELETED_SELF:
 *     return 'Deleted';
 * }
 * ```
 */
export const CustomerRole = {
  CUSTOMER: 'customer',
  COLONEL: 'colonel',
  ADMIN: 'admin',
  STAFF: 'staff',
  RECIPIENT: 'recipient',
  USER_DELETED_SELF: 'user_deleted_self',
  ANONYMOUS: 'anonymous',
} as const;

/**
 * Zod schema for validating customer role values.
 *
 * Strict — rejects anything outside the enum. Use for inputs the frontend
 * controls (forms, guards). For parsing backend records, use
 * {@link customerRoleResilientSchema} instead.
 *
 * @category Contracts
 */
export const customerRoleSchema = z.enum(customerRoleValues);

/**
 * Resilient role schema for parsing backend records.
 *
 * `.catch('customer')` over the bare enum: `role` is a display/UX field, and
 * an unrecognized value must degrade the role badge, never take down a whole
 * page. Before this, a role outside the enum failed the entire
 * AccountResponse parse and /account/settings/api rendered the error
 * boundary (#4298 / FRONTEND-19V — an account promoted to `admin` via the
 * CLI). The catch also absorbs legacy Redis hashes with a missing/null
 * `role`. Authorization is enforced server-side, so falling back to
 * 'customer' can only under-display, never over-grant.
 *
 * @category Contracts
 */
export const customerRoleResilientSchema = customerRoleSchema.catch('customer');

/**
 * Type guard for runtime customer role validation.
 *
 * @param role - String to validate
 * @returns True if role is a valid CustomerRole value
 *
 * @category Contracts
 * @example
 * ```typescript
 * const userInput = 'colonel';
 * if (isValidCustomerRole(userInput)) {
 *   // userInput is now typed as CustomerRole
 *   console.log(`Valid role: ${userInput}`);
 * }
 * ```
 */
export function isValidCustomerRole(role: string): role is CustomerRole {
  return customerRoleValues.includes(role as CustomerRole);
}

// ─────────────────────────────────────────────────────────────────────────────
// Feature flags
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Feature flags schema for customer-specific feature toggles.
 *
 * Allows any key-value pairs of boolean flags for flexible feature control.
 *
 * @category Contracts
 * @example
 * ```typescript
 * const customer = customerCanonical.parse(apiResponse);
 * if (customer.feature_flags['beta_features']) {
 *   // Feature is enabled for this customer
 * }
 * ```
 */
export const featureFlagsSchema = z.record(z.string(), z.boolean());

export type FeatureFlags = z.infer<typeof featureFlagsSchema>;

// ─────────────────────────────────────────────────────────────────────────────
// Customer canonical schema
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Canonical customer record contract.
 *
 * Defines field names and output types (post-parse).
 * No transforms - those are version-specific in shapes.
 *
 * Customer records track:
 * - Identity: objid (internal UUID), extid (external user-facing ID), email
 * - Status: role, verified, active
 * - Activity: secrets created/burned/shared, emails sent, last login
 * - Preferences: locale, notify_on_reveal
 * - Feature toggles: feature_flags
 *
 * @category Contracts
 * @see {@link "shapes/v2/customer".customerSchema} - V2 wire format
 * @see {@link "shapes/v3/customer".customerSchema} - V3 wire format
 *
 * @example
 * ```typescript
 * // Extend in version-specific shapes
 * const customerV3 = customerCanonical.extend({
 *   created: transforms.fromNumber.toDate,
 *   updated: transforms.fromNumber.toDate,
 * });
 *
 * // Derive TypeScript type
 * type Customer = z.infer<typeof customerCanonical>;
 * ```
 */
export const customerCanonical = z.object({
  // ─────────────────────────────────────────────────────────────────────────
  // Identity fields
  // ─────────────────────────────────────────────────────────────────────────

  /** Object ID (internal UUID, primary key). */
  objid: z.string(),

  /** External ID (user-facing, used in public APIs). */
  extid: z.string(),

  /** Email address (unique per customer). */
  email: z.email(),

  // ─────────────────────────────────────────────────────────────────────────
  // Status fields
  // ─────────────────────────────────────────────────────────────────────────

  /** User role determining authorization level. Unknown values degrade to
   * 'customer' rather than failing the record parse — see
   * customerRoleResilientSchema. */
  role: customerRoleResilientSchema,

  /** Whether email address has been verified. */
  verified: z.boolean(),

  /** Whether account is active (verified + role === customer). */
  active: z.boolean(),

  /** Whether user is a contributor (optional). */
  contributor: z.boolean().optional(),

  // ─────────────────────────────────────────────────────────────────────────
  // Activity counters
  // ─────────────────────────────────────────────────────────────────────────

  /** Number of secrets created by this customer. */
  secrets_created: z.number(),

  /** Number of secrets burned (destroyed) by this customer. */
  secrets_burned: z.number(),

  /** Number of secrets shared by this customer. */
  secrets_shared: z.number(),

  /** Number of notification emails sent for this customer. */
  emails_sent: z.number(),

  // ─────────────────────────────────────────────────────────────────────────
  // Timestamps
  // ─────────────────────────────────────────────────────────────────────────

  /** Last login timestamp (null if never logged in). */
  last_login: z.date().nullable(),

  /** Account creation timestamp. */
  created: z.date(),

  /** Last update timestamp. */
  updated: z.date(),

  // ─────────────────────────────────────────────────────────────────────────
  // Preferences
  // ─────────────────────────────────────────────────────────────────────────

  /** User locale preference (e.g., 'en', 'de'). */
  locale: z.string().nullable(),

  /** Whether to notify when a secret is revealed. */
  notify_on_reveal: z.boolean(),

  // ─────────────────────────────────────────────────────────────────────────
  // Feature flags
  // ─────────────────────────────────────────────────────────────────────────

  /** Customer-specific feature toggles. */
  feature_flags: featureFlagsSchema,
});

// ─────────────────────────────────────────────────────────────────────────────
// Type exports
// ─────────────────────────────────────────────────────────────────────────────

/** TypeScript type for customer record. */
export type CustomerCanonical = z.infer<typeof customerCanonical>;
